File: mclauncher/version_settings.py
# -*- coding: utf-8 -*-
"""每版本设置：隔离、内存、Java、JVM、启动前后命令、直连服务器。对齐 PCL 版本设置。"""
from __future__ import annotations

import os
import subprocess
from pathlib import Path

from . import utils
from .config import CONFIG

FILE_NAME = "pymcl.json"
ISOLATION_NONE = "none"
ISOLATION_SAVES = "saves"
ISOLATION_MODS = "mods"
ISOLATION_ALL = "all"
ISOLATION_LABELS = {
    ISOLATION_NONE: "大锅饭（与其他版本共用）",
    ISOLATION_SAVES: "隔离存档",
    ISOLATION_MODS: "隔离 Mod 与配置",
    ISOLATION_ALL: "完全独立",
}
SHARED_LINKS = ("mods", "config", "resourcepacks", "shaderpacks", "downloads")
SAVES_LINKS = ("saves",)
# 版本卡上的一键开关只在这两档之间翻，四档细分留在版本设置里
ISOLATED_DEFAULT = ISOLATION_ALL
# 转独立时从大锅饭里带一份过去的内容
SEED_DIRS = ("mods", "config", "resourcepacks", "shaderpacks")

DEFAULTS = {
    "isolation": ISOLATION_NONE,
    "memory_mb": None,
    "java": "自动选择",
    "jvm_args": "",
    "game_args": "",
    "pre_launch": "",
    "post_launch": "",
    "pre_launch_wait": True,
    "server": "",
    "port": "",
    "process_priority": "normal",
    "icon": "",
    "hidden": False,
    "login_account": "",
    "auth_server": "",
    "auth_server_name": "",
    "nide8_id": "",
    "gc": "",
    "window_title": "",
    "window_mode": "window",
    "window_width": None,
    "window_height": None,
    "skip_assets": False,
    "offline_skin": "default",
}

# UI 历史上写过 "maximize"，启动链早期只认 "fullscreen"，两边对不上导致全屏静默失效。
# 以 "maximize" 为准，另一个作为别名容错。
FULLSCREEN_MODES = ("maximize", "fullscreen")


def _file(instance, version_id) -> Path:
    return instance.versions_dir() / version_id / FILE_NAME


def load(instance, version_id) -> dict:
    data = dict(DEFAULTS)
    stored = utils.read_json(_file(instance, version_id), None)
    if isinstance(stored, dict):
        data.update(stored)
    iso = data.get("isolation") or CONFIG.get("default_isolation") or ISOLATION_NONE
    if iso not in ISOLATION_LABELS:
        iso = ISOLATION_NONE
    data["isolation"] = iso
    return data


def save(instance, version_id, data: dict) -> dict:
    cur = load(instance, version_id)
    cur.update(data or {})
    if cur.get("isolation") not in ISOLATION_LABELS:
        cur["isolation"] = ISOLATION_NONE
    utils.write_json(_file(instance, version_id), cur)
    return cur


def is_isolated(settings) -> bool:
    """这个版本是否有自己的 mods 目录（而不是吃大锅饭）。"""
    iso = (settings or {}).get("isolation") or ISOLATION_NONE
    return iso in (ISOLATION_MODS, ISOLATION_ALL)


def set_isolation(instance, version_id, mode, seed=False) -> dict:
    """切隔离模式。`seed=True` 时把大锅饭里现有的模组/配置复制一份过去。

    转独立那一下版本目录是空的，不带种子的话用户会以为模组丢了；
    但整合包版本转独立时又不该把别人的模组拖进来，所以交给调用方决定。
    """
    if mode not in ISOLATION_LABELS:
        raise ValueError(f"未知的隔离模式: {mode!r}")
    before = load(instance, version_id)
    data = save(instance, version_id, {"isolation": mode})
    if seed and is_isolated(data) and not is_isolated(before):
        _seed_from_shared(instance, version_id)
    apply_isolation(instance, version_id, data)
    return data


def _seed_from_shared(instance, version_id):
    import shutil

    src_root = Path(instance.path)
    dest_root = instance.versions_dir() / version_id
    for name in SEED_DIRS:
        src = src_root / name
        if not src.is_dir():
            continue
        dest = dest_root / name
        utils.ensure_dir(dest)
        for child in src.iterdir():
            target = dest / child.name
            if target.exists() or child.is_symlink():
                continue
            try:
                if child.is_dir():
                    shutil.copytree(child, target, dirs_exist_ok=True)
                else:
                    shutil.copy2(child, target)
            except OSError:
                continue


def game_dir(instance, version_id, settings=None) -> Path:
    settings = settings or load(instance, version_id)
    iso = settings.get("isolation") or ISOLATION_NONE
    if iso in (ISOLATION_ALL, ISOLATION_SAVES, ISOLATION_MODS):
        return instance.versions_dir() / version_id
    return Path(instance.path)


def mods_dir(instance, version_id, settings=None) -> Path:
    root = game_dir(instance, version_id, settings)
    return root / "mods"


def _junction(link: Path, target: Path):
    target = Path(target)
    link = Path(link)
    if link.exists() or link.is_symlink():
        if link.is_dir() and not link.is_symlink() and any(link.iterdir()):
            return
        try:
            if link.is_symlink() or link.is_file():
                link.unlink()
            elif link.is_dir() and not any(link.iterdir()):
                link.rmdir()
        except OSError:
            return
    utils.ensure_dir(target)
    utils.ensure_dir(link.parent)
    if os.name == "nt":
        subprocess.run(
            ["cmd", "/c", "mklink", "/J", str(link), str(target)],
            capture_output=True, check=False,
            creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0),
        )
        return
    try:
        link.symlink_to(target, target_is_directory=True)
    except OSError:
        pass


def apply_isolation(instance, version_id, settings=None) -> Path:
    """按隔离模式准备游戏目录。返回 game_dir。"""
    settings = settings or load(instance, version_id)
    gdir = game_dir(instance, version_id, settings)
    utils.ensure_dir(gdir)
    iso = settings.get("isolation") or ISOLATION_NONE
    if iso == ISOLATION_SAVES:
        for name in SHARED_LINKS:
            _junction(gdir / name, Path(instance.path) / name)
        utils.ensure_dir(gdir / "saves")
    elif iso == ISOLATION_MODS:
        for name in SAVES_LINKS + ("resourcepacks", "shaderpacks", "screenshots"):
            _junction(gdir / name, Path(instance.path) / name)
        for name in ("mods", "config"):
            utils.ensure_dir(gdir / name)
    elif iso == ISOLATION_ALL:
        # 从「隔离存档 / 隔离 Mod」切过来时，之前建的联接还指着共享池，
        # 不摘掉的话「完全独立」写进去的东西照样落在大锅饭里。
        for name in set(SHARED_LINKS) | set(SAVES_LINKS) | {"screenshots"}:
            _drop_link(gdir / name)
        for name in ("mods", "config", "saves", "resourcepacks", "shaderpacks"):
            utils.ensure_dir(gdir / name)
    return gdir


def _drop_link(path: Path):
    from .single_root import drop_link

    drop_link(path)

File: android/app/src/main/java/com/pymcl/mobile/data/VersionSettings.kt
package com.pymcl.mobile.data

import org.json.JSONObject
import java.io.File

/** 单个版本的设置，落在 `versions/<id>/pymcl.json`，字段与桌面 `version_settings.py` 同名同义。 */
data class VersionSetting(
    val isolation: String = VersionSettings.NONE,
    val memoryMb: Int? = null,
    val java: String = "自动选择",
    val jvmArgs: String = "",
    val gameArgs: String = "",
    val preLaunch: String = "",
    val postLaunch: String = "",
    val preLaunchWait: Boolean = true,
    val server: String = "",
    val port: String = "",
    val icon: String = "",
    val hidden: Boolean = false,
    val windowMode: String = "window",
    val windowTitle: String = "",
    val windowWidth: Int? = null,
    val windowHeight: Int? = null,
    val skipAssets: Boolean = false,
    /** GC 预设键，见 [GcPresets]；空串 = 跟随全局。 */
    val gc: String = "",
    val processPriority: String = "normal",
    /** 绑定到这个版本的账号名；空 = 跟随启动页。 */
    val loginAccount: String = "",
    val authServer: String = "",
    val authServerName: String = "",
    /** 统一通行证（Nide8）服务器 ID。 */
    val nide8Id: String = "",
    val offlineSkin: String = "default",
)

object VersionSettings {
    const val FILE_NAME = "pymcl.json"

    const val NONE = "none"
    const val SAVES = "saves"
    const val MODS = "mods"
    const val ALL = "all"

    /** 版本卡上那个一键开关只在「大锅饭」和这一档之间翻，四档细分留在版本设置里。 */
    const val ISOLATED_DEFAULT = ALL

    val LABELS = linkedMapOf(
        NONE to "大锅饭（与其他版本共用）",
        SAVES to "隔离存档",
        MODS to "隔离 Mod 与配置",
        ALL to "完全独立",
    )

    /** 转独立时从大锅饭里带一份过去的目录。 */
    val SEED_DIRS = listOf("mods", "config", "resourcepacks", "shaderpacks")

    fun file(instDir: File, version: String): File =
        File(instDir, "versions/$version/$FILE_NAME")

    fun load(instDir: File, version: String): VersionSetting {
        val f = file(instDir, version)
        if (!f.isFile) return VersionSetting()
        val obj = runCatching { JSONObject(f.readText()) }.getOrNull() ?: return VersionSetting()
        return fromJson(obj)
    }

    fun fromJson(obj: JSONObject): VersionSetting {
        val iso = obj.optString("isolation", NONE)
        return VersionSetting(
            isolation = if (iso in LABELS) iso else NONE,
            memoryMb = obj.optInt("memory_mb", 0).takeIf { it > 0 },
            java = obj.optString("java", "自动选择"),
            jvmArgs = obj.optString("jvm_args", ""),
            gameArgs = obj.optString("game_args", ""),
            preLaunch = obj.optString("pre_launch", ""),
            postLaunch = obj.optString("post_launch", ""),
            preLaunchWait = obj.optBoolean("pre_launch_wait", true),
            server = obj.optString("server", ""),
            port = obj.optString("port", ""),
            icon = obj.optString("icon", ""),
            hidden = obj.optBoolean("hidden", false),
            windowMode = obj.optString("window_mode", "window"),
            windowTitle = obj.optString("window_title", ""),
            windowWidth = obj.optInt("window_width", 0).takeIf { it > 0 },
            windowHeight = obj.optInt("window_height", 0).takeIf { it > 0 },
            skipAssets = obj.optBoolean("skip_assets", false),
            gc = obj.optString("gc", ""),
            processPriority = obj.optString("process_priority", "normal"),
            loginAccount = obj.optString("login_account", ""),
            authServer = obj.optString("auth_server", ""),
            authServerName = obj.optString("auth_server_name", ""),
            nide8Id = obj.optString("nide8_id", ""),
            offlineSkin = obj.optString("offline_skin", "default"),
        )
    }

    fun toJson(s: VersionSetting): JSONObject = JSONObject()
        .put("isolation", s.isolation)
        .put("memory_mb", s.memoryMb ?: JSONObject.NULL)
        .put("java", s.java)
        .put("jvm_args", s.jvmArgs)
        .put("game_args", s.gameArgs)
        .put("pre_launch", s.preLaunch)
        .put("post_launch", s.postLaunch)
        .put("pre_launch_wait", s.preLaunchWait)
        .put("server", s.server)
        .put("port", s.port)
        .put("icon", s.icon)
        .put("hidden", s.hidden)
        .put("window_mode", s.windowMode)
        .put("window_title", s.windowTitle)
        .put("window_width", s.windowWidth ?: JSONObject.NULL)
        .put("window_height", s.windowHeight ?: JSONObject.NULL)
        .put("skip_assets", s.skipAssets)
        .put("gc", s.gc)
        .put("process_priority", s.processPriority)
        .put("login_account", s.loginAccount)
        .put("auth_server", s.authServer)
        .put("auth_server_name", s.authServerName)
        .put("nide8_id", s.nide8Id)
        .put("offline_skin", s.offlineSkin)

    fun save(instDir: File, version: String, s: VersionSetting): VersionSetting {
        val f = file(instDir, version)
        f.parentFile?.mkdirs()
        f.writeText(toJson(s).toString(2), Charsets.UTF_8)
        return s
    }

    /** 这个版本有没有自己的 mods 目录（而不是吃大锅饭）。 */
    fun isolatedMods(s: VersionSetting): Boolean = s.isolation == MODS || s.isolation == ALL

    fun isolatedSaves(s: VersionSetting): Boolean = s.isolation == SAVES || s.isolation == ALL

    /** 只要不是纯大锅饭，游戏工作目录就落在版本目录里。 */
    fun isolatedAnywhere(s: VersionSetting): Boolean = s.isolation != NONE

    fun gameDir(instDir: File, version: String, s: VersionSetting = load(instDir, version)): File =
        if (isolatedAnywhere(s)) File(instDir, "versions/$version") else instDir

    fun modsDir(instDir: File, version: String, s: VersionSetting = load(instDir, version)): File =
        File(if (isolatedMods(s)) File(instDir, "versions/$version") else instDir, "mods")

    fun savesDir(instDir: File, version: String, s: VersionSetting = load(instDir, version)): File =
        File(if (isolatedSaves(s)) File(instDir, "versions/$version") else instDir, "saves")

    /**
     * 切隔离档位。`seed=true` 时把大锅饭里现有的模组 / 配置复制一份过去——
     * 转独立那一下版本目录是空的，不带种子用户会以为模组丢了。
     */
    fun setIsolation(
        instDir: File,
        version: String,
        mode: String,
        seed: Boolean = false,
    ): VersionSetting {
        require(mode in LABELS) { "未知的隔离模式: $mode" }
        val before = load(instDir, version)
        val after = save(instDir, version, before.copy(isolation = mode))
        if (seed && isolatedMods(after) && !isolatedMods(before)) {
            seedFromShared(instDir, version)
        }
        applyLayout(instDir, version, after)
        return after
    }

    internal fun seedFromShared(instDir: File, version: String) {
        val dest = File(instDir, "versions/$version")
        for (name in SEED_DIRS) {
            val src = File(instDir, name)
            if (!src.isDirectory) continue
            val target = File(dest, name)
            target.mkdirs()
            src.listFiles()?.forEach { child ->
                val out = File(target, child.name)
                if (out.exists()) return@forEach
                runCatching { if (child.isDirectory) child.copyRecursively(out) else child.copyTo(out) }
            }
        }
    }

    /** 按档位把该有的目录建出来。安卓没有 junction，共享靠启动时把路径指回实例根。 */
    fun applyLayout(
        instDir: File,
        version: String,
        s: VersionSetting = load(instDir, version),
    ): File {
        val dir = gameDir(instDir, version, s)
        dir.mkdirs()
        when (s.isolation) {
            SAVES -> File(dir, "saves").mkdirs()
            MODS -> listOf("mods", "config").forEach { File(dir, it).mkdirs() }
            ALL -> listOf("mods", "config", "saves", "resourcepacks", "shaderpacks")
                .forEach { File(dir, it).mkdirs() }
        }
        return dir
    }

    fun addressOf(s: VersionSetting): String {
        val host = s.server.trim()
        if (host.isEmpty()) return ""
        val port = s.port.trim().toIntOrNull()
        if (port == null || port !in 1..65535) return host
        return "$host:$port"
    }

    /**
     * 把用户在设置里敲的一行参数切成 argv。支持成对的单双引号，
     * 不配对的引号按普通字符处理——总比整行被吞掉强。
     */
    fun splitArgs(raw: String): List<String> {
        val out = mutableListOf<String>()
        val buf = StringBuilder()
        var quote = ' '
        var has = false
        for (ch in raw) {
            when {
                quote != ' ' -> {
                    if (ch == quote) quote = ' ' else buf.append(ch)
                }
                ch == '"' || ch == '\'' -> {
                    quote = ch
                    has = true
                }
                ch.isWhitespace() -> {
                    if (has || buf.isNotEmpty()) out += buf.toString()
                    buf.setLength(0)
                    has = false
                }
                else -> buf.append(ch)
            }
        }
        if (has || buf.isNotEmpty()) out += buf.toString()
        return out
    }
}
